// angular-inport/dist/inview-item.directive.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var position_resolver_1 = require("./utils/position-resolver");
var InviewItemDirective = (function () {
    function InviewItemDirective(_element) {
        this._element = _element;
    }
    Object.defineProperty(InviewItemDirective.prototype, "data", {
        set: function (d) {
            this._data = d;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InviewItemDirective.prototype, "id", {
        set: function (_id) {
            this._id = _id;
        },
        enumerable: true,
        configurable: true
    });
    InviewItemDirective.prototype.ngOnInit = function () {
    };
    InviewItemDirective.prototype.getELementRect = function () {
        return position_resolver_1.PositionResolver.getBoundingClientRect(this._element.nativeElement);
    };
    InviewItemDirective.prototype.isVisible = function () {
        return position_resolver_1.PositionResolver.isVisible(this._element.nativeElement);
    };
    InviewItemDirective.prototype.getData = function () {
        return { id: this._id, data: this._data };
    };
    return InviewItemDirective;
}());
__decorate([
    core_1.Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], InviewItemDirective.prototype, "data", null);
__decorate([
    core_1.Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], InviewItemDirective.prototype, "id", null);
InviewItemDirective = __decorate([
    core_1.Directive({
        selector: '[in-view-item]'
    }),
    __metadata("design:paramtypes", [core_1.ElementRef])
], InviewItemDirective);
exports.InviewItemDirective = InviewItemDirective;
