// ng2-metadata/index.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { NgModule } from '@angular/core';
import { MetadataLoader, MetadataStaticLoader, MetadataService } from './src/metadata.service';
export * from './src/models/page-title-positioning';
export * from './src/metadata.service';
export function metadataLoaderFactory() {
    return new MetadataStaticLoader();
}
var MetadataModule = MetadataModule_1 = (function () {
    function MetadataModule() {
    }
    MetadataModule.forRoot = function (providedLoader) {
        if (providedLoader === void 0) { providedLoader = {
            provide: MetadataLoader,
            useFactory: (metadataLoaderFactory)
        }; }
        return {
            ngModule: MetadataModule_1,
            providers: [
                providedLoader,
                MetadataService
            ]
        };
    };
    return MetadataModule;
}());
MetadataModule = MetadataModule_1 = __decorate([
    NgModule()
], MetadataModule);
export { MetadataModule };
var MetadataModule_1;
